mono/metadata/android.c

```diff
@@ -0,0 +1,537 @@
+	if (!TimeZone_class)
+		g_error ("%s: Fatal error: Could not find java.util.TimeZone class!", __func__);
+
+	TimeZone_getDefault = (*env)->GetStaticMethodID (env, TimeZone_class, "getDefault", "()Ljava/util/TimeZone;");
+	if (!TimeZone_getDefault)
+		g_error ("%s: Fatal error: Could not find java.util.TimeZone.getDefault() method!", __func__);
+
+	TimeZone_getID = (*env)->GetMethodID (env, TimeZone_class, "getID", "()Ljava/lang/String;");
+	if (!TimeZone_getID)
+		g_error ("%s: Fatal error: Could not find java.util.TimeZone.getDefault() method!", __func__);
+
+	NetworkInterface_class = lref_to_gref (env, (*env)->FindClass (env, "java/net/NetworkInterface"));
+	if (!NetworkInterface_class)
+		g_error ("Fatal error: Could not find java.net.NetworkInterface class!");
+
+	NetworkInterface_getByName = (*env)->GetStaticMethodID (env, NetworkInterface_class, "getByName", "(Ljava/lang/String;)Ljava/net/NetworkInterface;");
+	if (!NetworkInterface_getByName)
+		g_error ("Fatal error: Could not find java.net.NetworkInterface.getByName() method!");
+
+	NetworkInterface_isUp = (*env)->GetMethodID (env, NetworkInterface_class, "isUp", "()Z");
+	if (!NetworkInterface_isUp)
+		g_error ("Fatal error: Could not find java.net.NetworkInterface.isUp() method!");
+
+	NetworkInterface_supportsMulticast = (*env)->GetMethodID (env, NetworkInterface_class, "supportsMulticast", "()Z");
+	if (!NetworkInterface_supportsMulticast)
+		g_error ("Fatal error: Could not find java.net.NetworkInterface.supportsMulticast() method!");
+
+	initialized = TRUE;
+}
+
+JNIEXPORT jint JNICALL
+JNI_OnLoad (JavaVM *vm, gpointer reserved)
+{
+	mono_jvm_initialize (vm);
+	return JNI_VERSION_1_6;
+}
+
+JNIEnv*
+mono_jvm_get_jnienv (void)
+{
+	JNIEnv *env;
+
+	g_assert (initialized);
+
+	(*jvm)->GetEnv (jvm, (void**)&env, JNI_VERSION_1_6);
+	if (env)
+		return env;
+
+	mono_thread_attach (mono_domain_get ());
+
+	(*jvm)->GetEnv (jvm, (void**)&env, JNI_VERSION_1_6);
+	if (env)
+		return env;
+
+	g_error ("%s: Fatal error: Could not create env", __func__);
+}
+
+struct BundledProperty {
+	gchar *name;
+	gchar *value;
+	gint   value_len;
+	struct BundledProperty *next;
+};
+
+static struct BundledProperty* bundled_properties;
+
+static struct BundledProperty*
+lookup_system_property (const gchar *name)
+{
+	struct BundledProperty *p = bundled_properties;
+	for ( ; p ; p = p->next)
+		if (strcmp (p->name, name) == 0)
+			return p;
+	return NULL;
+}
+
+void
+monodroid_add_system_property (const gchar *name, const gchar *value)
+{
+	gint name_len, value_len;
+
+	struct BundledProperty* p = lookup_system_property (name);
+	if (p) {
+		gchar *n = g_strdup (value);
+		g_free (p->value);
+		p->value      = n;
+		p->value_len  = strlen (p->value);
+		return;
+	}
+
+	name_len  = strlen (name);
+	value_len = strlen (value);
+
+	p = g_malloc0 (sizeof (struct BundledProperty) + name_len + 1);
+
+	p->name = ((char*) p) + sizeof (struct BundledProperty);
+	strncpy (p->name, name, name_len);
```

Comment:
This can be a more efficienct memcpy (p->name, name, name_len +1 ), given we just did the  allocation.